### PR TITLE
[CDAP-15738] Upgrades node version and corresponding node modules for compatibility

### PR DESCRIPTION
**Pull request by ajainarayanan**
JIRA: https://issues.cask.co/browse/CDAP-15738
Build: https://builds.cask.co/browse/CDAP-UDUT377-1

- Upgrades node version to 10.16.2
- Upgrades babel and gulp node modules for compatibility
- Fixes Gulpfile and webpack build config files based on upgrade

### CODE REVIEW

File: cdap-ui/server.js [hunk 1]
```diff
@@ -112,4 +111,21 @@ cdapConfigurator.getCDAPConfig()
...
+      const nodejsport = cdapConfig['dashboard.bind.port'];
+      const nodejsprotocol = (cdapConfig['ssl.external.enabled'] === 'true') ? 'https' : 'http';
+      let allowedOrigin;
+      if (nodejsport) {
+        allowedOrigin = `${nodejsprotocol}://${nodejsserver}:${nodejsport}`;
+      } else {
+        allowedOrigin = `${nodejsprotocol}://${nodejsserver}`;
+      }
+      if (req.headers.origin !== allowedOrigin) {
+        log.info('Unknown Orign: ' + req.headers.origin);
```
Comment (yukiej) [inline cdap-ui/server.js]: nit: typo in "origin"

**Review by yukiej — COMMENTED** [review]
lgtm!

**Comment by elfenheart** [comment]
The connection seemed to get rejected on Sandbox as well (probably due to localhost vs 127.0.0.1). However, even when the connection is rejected, I can still use the UI and do queries. Is that expected?

**Comment by ajainarayanan** [comment]
Talked to @elfenheart about it. Since we are using sockjs it resorts to use other modes of communication to polyfil websockets (xhr_polling). That is expected.

File: cdap-ui/server.js [hunk 2]
```diff
@@ -112,4 +111,21 @@ cdapConfigurator.getCDAPConfig()
     });
 
     sockServer.installHandlers(server, { prefix: '/_sock' });
+    server.addListener('upgrade', function (req, socket, head) {
+      const nodejsserver = cdapConfig['dashboard.bind.address'];
+      const nodejsport = cdapConfig['dashboard.bind.port'];
+      const nodejsprotocol = (cdapConfig['ssl.external.enabled'] === 'true') ? 'https' : 'http';
+      let allowedOrigin;
+      if (nodejsport) {
+        allowedOrigin = `${nodejsprotocol}://${nodejsserver}:${nodejsport}`;
```
Comment (elfenheart) [inline cdap-ui/server.js]: `127.0.0.1` and `localhost` should be interchangeable. Currently cdap-default for sandbox is bind to `127.0.0.1`

**Comment by ajainarayanan** [comment]
Currently on-hold as checking for origin alone will be problematic in some environments and is not the full solution. Working on adding a csrf-like token that will be validated on websocket upgrade or any http request.

**Comment by ajainarayanan** [comment]
Changes are up to date. Fixing tests to pass.

File: cdap-ui/package.json [hunk 3]
```diff
@@ -38,14 +38,20 @@
...
+    "@babel/plugin-proposal-class-properties": "^7.0.0",
+    "@babel/plugin-proposal-decorators": "^7.0.0",
+    "@babel/plugin-proposal-export-namespace-from": "^7.0.0",
+    "@babel/plugin-proposal-function-sent": "^7.0.0",
+    "@babel/plugin-proposal-json-strings": "^7.0.0",
+    "@babel/plugin-proposal-numeric-separator": "^7.0.0",
+    "@babel/plugin-proposal-object-rest-spread": "7.0.0",
+    "@babel/plugin-proposal-throw-expressions": "^7.0.0",
+    "@babel/plugin-syntax-dynamic-import": "^7.0.0",
+    "@babel/plugin-syntax-import-meta": "^7.0.0",
```
Comment (elfenheart) [inline cdap-ui/package.json]: nit: remove the `^` and redo yarn install

**Review by elfenheart — COMMENTED** [review]
I think you'll need to update pom.xml to change the node version as well

**Comment by yukiej** [comment]
Should the cdap-ui/BUILD.rst get updated as well, with the nodeJS version? https://github.com/cdapio/cdap/blob/develop/cdap-ui/BUILD.rst

**Review by yukiej — APPROVED** [review]
One question/comment, otherwise LGTM!

File: cdap-distributions/src/packer/files/cdap-sdk.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
   },
   "nodejs": {
     "install_method": "binary",
-    "version": "8.7.0",
+    "version": "10.16.2",
     "binary": {
       "checksum": {
         "linux_x64": "115c7bd133170fd7a1bf408b2e293021e4b5a80a66a4962829ce5d362ce43762"
```
Comment (dereklwood) [inline cdap-distributions/src/packer/files/cdap-sdk.json]: this needs updating

**Review by elfenheart — APPROVED** [review]
👍 LGTM

**Comment by ajainarayanan** [comment]
Three consecutive green build: [Build 1](https://builds.cask.co/browse/CDAP-UDUT377-13), [Build 2](https://builds.cask.co/build/result/gotoBuildResult.action?buildKey=CDAP-UDUT377&buildNumber=12&returnUrl=%2Fbrowse%2FCDAP-UDUT377-13), [Build 3](https://builds.cask.co/build/result/gotoBuildResult.action?buildKey=CDAP-UDUT377&buildNumber=11&returnUrl=%2Fbrowse%2FCDAP-UDUT377-13)

**Pull request merged**